oadp-operator/controllers/registry.go

```diff
@@ -0,0 +1,350 @@
+	backupStorageLocationList := velerov1.BackupStorageLocationList{}
+
+	// Fetch the configured backupstoragelocations
+	if err := r.List(r.Context, &backupStorageLocationList, bslListOptions); err != nil {
+		return false, err
+	}
+
+	// Loop through all the configured BSLs and create registry for each of them
+	for _, bsl := range backupStorageLocationList.Items {
+		registryDeployment := &appsv1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      constructBSLRegistryName(&bsl),
+				Namespace: VeleoNamespace,
+			},
+		}
+
+		op, err := controllerutil.CreateOrUpdate(r.Context, r.Client, registryDeployment, func() error {
+
+			// Setting Registry Deployment selector if a new object is created as it is immutable
+			if registryDeployment.ObjectMeta.CreationTimestamp.IsZero() {
+				registryDeployment.Spec.Selector = &metav1.LabelSelector{
+					MatchLabels: map[string]string{
+						"component": constructBSLRegistryName(&bsl),
+					},
+				}
+			}
+
+			// update the Registry Deployment template
+			err := r.buildRegistryDeployment(registryDeployment, &bsl)
+			return err
+		})
+
+		if err != nil {
+			return false, err
+		}
+
+		//TODO: Review registry deployment status and report errors and conditions
+
+		if op == controllerutil.OperationResultCreated || op == controllerutil.OperationResultUpdated {
+			// Trigger event to indicate registry deployment was created or updated
+			r.EventRecorder.Event(registryDeployment,
+				corev1.EventTypeNormal,
+				"RegistryDeploymentReconciled",
+				fmt.Sprintf("performed %s on registry deployment %s/%s", op, registryDeployment.Namespace, registryDeployment.Name),
+			)
+		}
+
+	}
+
+	return true, nil
+}
+
+// Construct and update the registry deployment for a bsl
+func (r *VeleroReconciler) buildRegistryDeployment(registryDeployment *appsv1.Deployment, bsl *velerov1.BackupStorageLocation) error {
+
+	// Setting controller owner reference on the registry deployment
+	err := controllerutil.SetControllerReference(bsl, registryDeployment, r.Scheme)
+	if err != nil {
+		return err
+	}
+
+	registryDeployment.Labels = r.getRegistryBSLLabels(bsl)
+
+	registryDeployment.Spec = appsv1.DeploymentSpec{
+		Replicas: pointer.Int32(1),
+		Template: corev1.PodTemplateSpec{
+			ObjectMeta: metav1.ObjectMeta{
+				Labels: map[string]string{
+					"component": constructBSLRegistryName(bsl),
+				},
+			},
+			Spec: corev1.PodSpec{
+				RestartPolicy: corev1.RestartPolicyAlways,
+				Containers:    r.buildRegistryContainer(bsl),
+			},
+		},
+	}
+	return nil
+}
+
+func (r *VeleroReconciler) getRegistryBSLLabels(bsl *velerov1.BackupStorageLocation) map[string]string {
+	labels := map[string]string{
+		"app.kubernetes.io/name":       OADPOperatorVelero,
+		"app.kubernetes.io/instance":   constructBSLRegistryName(bsl),
+		"app.kubernetes.io/managed-by": OADPOperator,
+		"app.kubernetes.io/component":  Registry,
+	}
+	return labels
+}
+
+func constructBSLRegistryName(bsl *velerov1.BackupStorageLocation) string {
```

Comment:
rename this to registryName()? 

Since it takes BSL as input, we know that it is fro BSL registry, since it returns string, we know it constructs something so dropping those two from name will not hurt IMO